pkg/driver/controller_modify_volume.go

```diff
@@ -17,8 +20,147 @@ const (
+		return fmt.Errorf("Different size was requested by a previous request")
+	}
+	if r.modifyDiskOptions.IOPS != 0 && h.mergedRequest.modifyDiskOptions.IOPS != 0 && r.modifyDiskOptions.IOPS != h.mergedRequest.modifyDiskOptions.IOPS {
+		return fmt.Errorf("Different IOPS was requested by a previous request")
+	}
+	if r.modifyDiskOptions.Throughput != 0 && h.mergedRequest.modifyDiskOptions.Throughput != 0 && r.modifyDiskOptions.Throughput != h.mergedRequest.modifyDiskOptions.Throughput {
+		return fmt.Errorf("Different throughput was requested by a previous request")
+	}
+	if r.modifyDiskOptions.VolumeType != "" && h.mergedRequest.modifyDiskOptions.VolumeType != "" && r.modifyDiskOptions.VolumeType != h.mergedRequest.modifyDiskOptions.VolumeType {
+		return fmt.Errorf("Different volume type was requested by a previous request")
+	}
+	return nil
+}
+
+func (h *modifyVolumeRequestHandler) mergeModifyVolumeRequest(r *modifyVolumeRequest) {
+	h.mux.Lock()
+	defer h.mux.Unlock()
+	if r.newSize != 0 {
+		h.mergedRequest.newSize = r.newSize
+	}
+	if r.modifyDiskOptions.IOPS != 0 {
+		h.mergedRequest.modifyDiskOptions.IOPS = r.modifyDiskOptions.IOPS
+	}
+	if r.modifyDiskOptions.Throughput != 0 {
+		h.mergedRequest.modifyDiskOptions.Throughput = r.modifyDiskOptions.Throughput
+	}
+	if r.modifyDiskOptions.VolumeType != "" {
+		h.mergedRequest.modifyDiskOptions.VolumeType = r.modifyDiskOptions.VolumeType
+	}
+}
+
+func (d *controllerService) addModifyVolumeRequest(ctx context.Context, volumeID string, r *modifyVolumeRequest) error {
+	d.modifyVolumeManager.mux.RLock()
+	requestHandler, ok := d.modifyVolumeManager.requestHandlerMap[volumeID]
+	d.modifyVolumeManager.mux.RUnlock()
+	if ok {
+		err := requestHandler.validateModifyVolumeRequest(r)
+		if err == nil {
+			requestHandler.requestChan <- r
+		} else {
+			return err
+		}
+	} else {
+		d.modifyVolumeManager.mux.Lock()
+		defer d.modifyVolumeManager.mux.Unlock()
+
+		requestChan := make(chan *modifyVolumeRequest)
+		errChan := make(chan error)
+		resultChan := make(chan int64)
+		requestHandler := modifyVolumeRequestHandler{
+			requestChan:   requestChan,
+			errChan:       errChan,
+			resultChan:    resultChan,
+			mergedRequest: r,
+			volumeID:      volumeID,
+			mux:           sync.RWMutex{},
+		}
+		d.modifyVolumeManager.requestHandlerMap[volumeID] = &requestHandler
+
+		go d.startVolumeTimer(ctx, &requestHandler)
+	}
+	return nil
+}
+
+func (d *controllerService) startVolumeTimer(ctx context.Context, h *modifyVolumeRequestHandler) {
+	klog.V(4).InfoS("Starting ModifyVolume timer for ", "volumeID", h.volumeID)
+	for {
+		select {
+		case req := <-h.requestChan:
+			h.mergeModifyVolumeRequest(req)
+		case <-time.After(modifyVolumeRequestHandlerTimeout):
+			actualSizeGiB, err := d.executeModifyVolumeRequest(ctx, h.volumeID, h.mergedRequest)
+			if err != nil {
+				h.errChan <- err
+			} else {
+				h.resultChan <- actualSizeGiB
+			}
+			return
+		case <-ctx.Done():
```

Comment:
So I've been thinking about this for awhile, and I'm pretty sure we need to remove this (this function probably shouldn't accept a context at all).

In the event that this context gets cancelled, the coalescing will permanently deadlock for that volume, because `startVolumeTimer` is only ever called a single time for any given coalescing attempt. Thus, existing because the context of the first request times out will permanently deadlock the volume, as no new instance of `startVolumeTimer` will ever be created.

Instead, context should be handled in the same place as waiting on the results/error channels (see my comment there).

Reply:
``startVolumeTimer`` needs to accept ``context`` because ``cloud.ResizeOrModifyDisk`` needs ``context``.

Reply:
and yes we should remove this case on ``ctx.Done()`` here and have it handled in ``controller.go``.